### PR TITLE
Provide support for categorical variables in optimized CPU inference

### LINKED ISSUE #207: Add categorical feature support to Herring

### PR DESCRIPTION
**Pull request by wphicks**
Resolve #207

### CODE REVIEW

File: src/herring/tl_helpers.hpp [hunk 1]
```diff
@@ -151,10 +153,36 @@ auto convert_tree(treelite::Tree<tl_threshold_t, tl_output_t> const& tl_tree, bo
...
+        } else {
+          hot_child = right_id;
+          distant_child = left_id;
+        }
+        auto tl_categories = tl_tree.MatchingCategories(cur_node_id);
+        auto constexpr max_category = typename tree_t::node_type::category_set_type{}.size();
+        auto bit_representation = std::transform_reduce(
+          std::begin(tl_categories),
+          std::end(tl_categories),
+          std::size_t{},
```
Comment (hcho3) [inline src/herring/tl_helpers.hpp:168]: Could it be possible for `1 << max_category` to be too large to fit in `std::size_t`?

**Review by hcho3 — COMMENTED** [review]
Overall LGTM. I see a fair bit of duplicated code in `simple_tree` and `lookup_tree`, all related to `evaluate_tree_node`. I suppose this duplication is better than other possible alternatives (inheritance, template metaprogramming)?

**Comment by wphicks** [comment]
> I see a fair bit of duplicated code in simple_tree and lookup_tree
I agree 100%. Template metaprogramming resolves it _fairly_ cleanly, but as I started to do it, I realized that there is an even nicer way we can set this up with a more significant refactor. I decided to save that for a separate PR to get categorical in now, and then I'll follow up with something that should clean up the whole codebase a fair amount. Seem all right to you?

File: src/herring/tl_helpers.hpp [hunk 2]
```diff
@@ -151,10 +153,25 @@ auto convert_tree(treelite::Tree<tl_threshold_t, tl_output_t> const& tl_tree, bo
...
-        throw unconvertible_model_exception{"Categorical nodes not yet implemented"};
+        if (tl_tree.CategoriesListRightChild(cur_node_id)) {
+          hot_child = left_id;
+          distant_child = right_id;
+        } else {
+          hot_child = right_id;
+          distant_child = left_id;
+        }
+        auto tl_categories = tl_tree.MatchingCategories(cur_node_id);
+        auto constexpr max_category = typename tree_t::node_type::category_set_type{}.size();
```
Comment (hcho3) [inline src/herring/tl_helpers.hpp:164]: Is this used?

Comment (wphicks) [reply src/herring/tl_helpers.hpp:164]: Oh darn! Thank you for catching that. It's not, but it should be. Just pushed an update.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**